sessiontxn/staleread/util.go

```diff
@@ -65,3 +66,23 @@ func CalculateTsWithReadStaleness(sctx sessionctx.Context, readStaleness time.Du
 func IsStmtStaleness(sctx sessionctx.Context) bool {
 	return sctx.GetSessionVars().StmtCtx.IsStaleness
 }
+
+// CalculateTsWithExternalTimestamp returns the minimum tso of `minSafeTs` and `externalTimestamp` variable
+func CalculateTsWithExternalTimestamp(ctx context.Context, sctx sessionctx.Context) (uint64, error) {
+	externalTimestamp, err := getExternalTimestamp(ctx, sctx)
+	if err != nil {
+		return 0, errAsOf.FastGenWithCause(err.Error())
+	}
+	return calculateTsWithExternalTimestamp(sctx, externalTimestamp), nil
+}
+
+// calculateTsWithExternalTimestamp returns the minimum tso of `minSafeTs` and `externalTimestamp`
+func calculateTsWithExternalTimestamp(sctx sessionctx.Context, externalTimestamp uint64) uint64 {
+	minSafeTs := expression.GetMinSafeTS(sctx)
+
+	if minSafeTs != 0 && minSafeTs < externalTimestamp {
+		return minSafeTs
```

Comment:
Should we return an error here? Maybe the result is unexpected

Reply:
Yep! I'll use the `externalTimestamp` directly, the behavior is much clearer. The tradeoff is that the execution could fail and will need retry multiple times waiting for the `safe-ts` to promote.

Reply:
Fixed.